jujugui/static/gui/src/app/app.js

```diff
@@ -1654,12 +1657,14 @@ YUI.add('juju-gui', function(Y) {
       }
 
       var socketUrl = this.env.get('socket_url');
-      // XXX frankban: this is not generic, and very specific for how the
-      // socket URL is composed in the GUI embedded in OpenStack. Therefore,
-      // the logic here for calculating the new socket URL for the given UUID
-      // must be considered temporary demo code.
-      var baseUrl = socketUrl.substring(0, socketUrl.lastIndexOf('/'));
-      var newSocketUrl = baseUrl + '/' + uuid;
+      var newSocketUrl, baseUrl;
```

Comment:
I guess this is ok for now, but it will likely need to be reworked later.
The switchEnv method does not yet support multiple controllers (it does not receive an address).
The embedded flag semantic is more similar to embeddedInHorizon than embeddedInGeneral, as we don't know about available URLs in public clouds. What I am thinking is something in which we pass a template for the WebSocket URL in config. Like `config.socketTemplate = '/juju/api/$address/$port/$uuid'` or `config.socketTemplate = '/ws/environment/$uuid/api'`. In both cases those are paths of 'wss://$currenthost:$currentport`, which can be easily retrieved. In config, the GUI takes the socket template and the initial URL (either as the current socket URL or as separate options like socketAddress and socketUUID). The switchEnv is changed to receive a "host:port" address. At that point you can use host, port and uuid as a context to render the socket URL based on the socketTemplate in a generic way, and the embedded is no longer required. This would also solve the current issue in _generateSocketUrl: if there is a jem, then /juju/api is always used, but a jem URL could be provided even if we are not embedded in horizon. That function needs to change as well, and having a separate renderSocketUrl(template, host, port, uuid) would be great. What do you think? I am not saying this must be done now, and what we have here is surely a step forward. But if you agree with my plan, please let's add at least an XXX describing it.